src/coverlet.core/Abstracts/IInstrumenter.cs

```diff
@@ -0,0 +1,42 @@
+﻿#nullable enable
+
+using System.IO;
+using System.Reflection;
+
+using Coverlet.Core.ObjectModel;
+
+namespace Coverlet.Core.Abstracts
+{
+    public class InstrumentationOptions
+    {
+        public string? Module { get; set; }
+        public string[]? IncludeFilters { get; set; }
+        public string[]? ExcludeFilters { get; set; }
+        public string[]? ExcludeSourceFiles { get; set; }
+        public string[]? ExcludeAttributes { get; set; }
+        public string[]? IncludeDirectories { get; set; }
+        public bool IncludeTestAssembly { get; set; }
+        public bool SingleHit { get; set; }
+        public bool UseSourceLink { get; set; }
+        public string? MergeWith { get; set; }
+    }
+
+    public interface ICoverageEngineFactory
+    {
+        ICoverageEngine CreateEngine(InstrumentationOptions options);
+        IReporter CreateReporter(string format);
+        IInProcessCoverageEngine CreateInProcessEngine(Stream instrumentationResultStream);
+    }
+
+    public interface ICoverageEngine
```

Comment:
Could this be IDisposable instead and clean up the instrumented files on dispose? It would let the calling code control when it happens instead of it being done on process exit.

Reply:
> clean up the instrumented files

You mean "restore non instrumented ones"?

Reply:
Yes :)

Reply:
Could be possible...I need to check how much code update...but I think that we could inject a slighly different instrumenterhelper that don't fire restore on process exits, I'll take a look, thank's.

Reply:
The latter is also a good tool to look at instrumented files themselves.  Could be an enum on `InstrumenterOptions` when calling `ICoverageEngineFactory.CreateEngine`:

    enum RestoreInstrumentedModules
    {
           OnProcessExit,
           OnEngineDispose,
           Never
     }

It would mean that `ICoverageEngine` still need to be `IDisposable`, even if it is only cleans up resources in some cases.

Reply:
I'll try to add this feat.